pkg/adapter/common/backoff.go

```diff
@@ -0,0 +1,107 @@
+/*
+Copyright (c) 2020 TriggerMesh Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package common
+
+import (
+	"context"
+	"math"
+	"time"
+)
+
+// default values for backoff
+const (
+	expFactor = 2
+
+	minBackoff = 1 * time.Second
+	maxBackoff = 32 * time.Second
+)
+
+// Backoff provides a simple exponential backoff mechanism
+type Backoff struct {
+	step     int
+	factor   float64
+	min, max time.Duration
+}
+
+// NewBackoff accepts optional values for minimum and maximum wait period
+// and return new instance of Backoff structure
+func NewBackoff(args ...time.Duration) *Backoff {
+	backoff := Backoff{
+		step:   0,
+		factor: expFactor,
+		min:    minBackoff,
+		max:    maxBackoff,
+	}
+
+	switch len(args) {
+	case 1:
+		backoff.min = args[0]
+	case 2:
+		backoff.min = args[0]
+		backoff.max = args[1]
```

Comment:
We should assert that `backoff.min <= backoff.max` and just ignore if it's not the case, maybe along with a small test.